packages/blocks/plus_minus/src/index.js

```diff
@@ -0,0 +1,1048 @@
+   * @param {number} targetCount The number of inputs the block should have.
+   * @this Blockly.Block
+   * @private
+   */
+  updateShape_: function(targetCount) {
+    while (this.itemCount_ < targetCount) {
+      this.addPart_();
+    }
+    while (this.itemCount_ > targetCount) {
+      this.removePart_();
+    }
+    this.updateMinus_();
+  },
+
+  /**
+   * Callback for the plus image. Adds an input to the block and updates the
+   * state of the minus.
+   * @this Blockly.Block
+   */
+  plus: function() {
+    this.addPart_();
+    this.updateMinus_();
+  },
+
+  /**
+   * Callback for the minus image. Removes the input at the end of the block and
+   * updates the state of the minus.
+   * @this Blockly.Block
+   */
+  minus: function() {
+    if (this.itemCount_ == 0) {
+      return;
+    }
+    this.removePart_();
+    this.updateMinus_();
+  },
+
+  // To properly keep track of indices we have to increment before/after adding
+  // the inputs, and decrement the opposite.
+  // Because we want our first item to be ADD0 (not ADD1) we increment after.
+
+  /**
+   * Adds an input to the end of the block. If the block currently has no
+   * inputs it updates the top 'EMPTY' input to receive a block.
+   * @this Blockly.Block
+   * @private
+   */
+  addPart_: function() {
+    if (this.itemCount_ == 0) {
+      if (this.getInput('EMPTY')) {
+        // TODO: I don't think this should throw errors. It would be nice if
+        //   It returned a boolean instead.
+        this.removeInput('EMPTY');
+      }
+      this.topInput_ = this.appendValueInput('ADD' + this.itemCount_)
+          .appendField(new FieldPlus(), 'PLUS')
+          .appendField(Blockly.Msg['TEXT_JOIN_TITLE_CREATEWITH']);
+    } else {
+      this.appendValueInput('ADD' + this.itemCount_);
+    }
+    this.itemCount_++;
+  },
+
+  /**
+   * Removes an input from the end of the block. If we are removing the last
+   * input this updates the block to have an 'EMPTY' top input.
+   * @this Blockly.Block
+   * @private
+   */
+  removePart_: function() {
+    this.itemCount_--;
+    this.removeInput('ADD' + this.itemCount_);
+    if (this.itemCount_ == 0) {
+      this.topInput_ = this.appendDummyInput('EMPTY')
+          .appendField(new FieldPlus(), 'PLUS')
+          .appendField(this.newQuote_(true))
+          .appendField(this.newQuote_(false));
+    }
+  },
+
+  /**
+   * Makes it so the minus is visible iff there is an input available to remove.
+   * @private
+   */
+  updateMinus_: function() {
+    const minusField = this.getField('MINUS');
+    if (!minusField && this.itemCount_ > 0) {
+      this.topInput_.insertFieldAt(1, new FieldMinus(), 'MINUS');
+    } else if (minusField && this.itemCount_ < 1) {
+      this.topInput_.removeField('MINUS');
+    }
+  },
+};
+
+/**
+ * Adds the quotes mixin to the block. Also updates the shape so that if no
+ * mutator is provided the block has two inputs.
+ * @this Blockly.Block
+ * @constructor
```

Comment:
Why the constructor annotation?  It's called in the constructor, but it's not a constructor fn on its own.